src/test/java/org/terracotta/statistics/simulation/latency/ConcurrentParameterized.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * All content copyright Terracotta, Inc., unless otherwise indicated.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.statistics.simulation.latency;
+
+import org.junit.runners.Parameterized;
+import org.junit.runners.ParentRunner;
+import org.junit.runners.model.RunnerScheduler;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Enables to run {@link Parameterized} Junit's test in parallel
+ *
+ * @author Mathieu Carbou
+ */
+public class ConcurrentParameterized extends Parameterized {
```

Comment:
I've tried. Let's say you do that:

in ctor:

```
    setScheduler(new RunnerScheduler() {
      private final ExecutorService service = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());

      public void schedule(Runnable childStatement) {
        service.submit(childStatement);
      }

      public void finished() {
        try {
          service.shutdown();
          service.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
        } catch (InterruptedException e) {
          e.printStackTrace(System.err);
        }
      }
    });
```

and you override:


  @Override
  public void setScheduler(RunnerScheduler scheduler) {
    for (Runner child : getChildren()) {
      if (child instanceof ParentRunner<?>) {
        ((ParentRunner<?>) child).setScheduler(scheduler);
      }
    }
  }

Then there are 2 mistakes:

1. You might submitting to the executor more jobs than it can execute. If you try you'll get this error:

```
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@278bb07e rejected from java.util.concurrent.ThreadPoolExecutor@67d48005[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
```

2. If you just move to a cached pool, then you might create a LOT of threads

3. The `@AfterClass` will be executed before the end of the tests, because there is no ordering between test instances and the class runners

What I am doing with my code is that I am only using 1 executor, and I am submitting only all test instances into the executor, but not the principal runner, so that BeforeClass and AfterClass are executed before and after all tests ran concurrently, and also I am making sure to only close the executor once when the principal finishes.

Let me know if you find a shorter / smarter way to do that... I've tried the solution proposed her (https://sqa.stackexchange.com/questions/6982/junit-parallelcomputer-runs-all-test-classes-as-one/7498) but it does not work when running directly a Parameterized test.